File: apps/user-app-android/app/src/main/java/com/parkgolf/app/presentation/feature/notifications/NotificationsViewModel.kt
package com.parkgolf.app.presentation.feature.notifications

import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CalendarMonth
import androidx.compose.material.icons.filled.Notifications
import androidx.compose.material.icons.filled.People
import androidx.compose.material.icons.filled.Shield
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.parkgolf.app.domain.model.AppNotification
import com.parkgolf.app.domain.model.NotificationType
import com.parkgolf.app.domain.repository.NotificationRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

// Filter enum for notifications
enum class NotificationFilter(
    val label: String,
    val icon: ImageVector,
    val types: List<NotificationType>?
) {
    ALL(
        label = "전체",
        icon = Icons.Default.Notifications,
        types = null
    ),
    BOOKING(
        label = "예약",
        icon = Icons.Default.CalendarMonth,
        types = listOf(
            NotificationType.BOOKING_CONFIRMED,
            NotificationType.BOOKING_CANCELLED,
            NotificationType.PAYMENT_SUCCESS,
            NotificationType.PAYMENT_FAILED
        )
    ),
    SOCIAL(
        label = "소셜",
        icon = Icons.Default.People,
        types = listOf(
            NotificationType.FRIEND_REQUEST,
            NotificationType.FRIEND_ACCEPTED,
            NotificationType.CHAT_MESSAGE
        )
    ),
    SYSTEM(
        label = "시스템",
        icon = Icons.Default.Shield,
        types = listOf(NotificationType.SYSTEM_ALERT)
    );

    fun matches(type: NotificationType): Boolean {
        return types == null || types.contains(type)
    }

    fun emptyMessage(): Pair<String, String> = when (this) {
        ALL -> "알림이 없습니다" to "새로운 알림이 도착하면 여기에 표시됩니다"
        BOOKING -> "예약 알림이 없습니다" to "예약 관련 알림이 도착하면 여기에 표시됩니다"
        SOCIAL -> "소셜 알림이 없습니다" to "친구 요청이나 채팅 알림이 도착하면 여기에 표시됩니다"
        SYSTEM -> "시스템 알림이 없습니다" to "시스템 공지사항이 도착하면 여기에 표시됩니다"
    }
}

data class NotificationsUiState(
    val isLoading: Boolean = false,
    val isLoadingMore: Boolean = false,
    val notifications: List<AppNotification> = emptyList(),
    val selectedFilter: NotificationFilter = NotificationFilter.ALL,
    val error: String? = null,
    val currentPage: Int = 1,
    val totalPages: Int = 1
) {
    val hasMorePages: Boolean
        get() = currentPage < totalPages

    val filteredNotifications: List<AppNotification>
        get() {
            val types = selectedFilter.types ?: return notifications
            return notifications.filter { types.contains(it.type) }
        }

    val unreadCount: Int
        get() = notifications.count { !it.isRead }

    val unreadCounts: Map<NotificationFilter, Int>
        get() {
            return NotificationFilter.entries.associateWith { filter ->
                val types = filter.types
                if (types == null) {
                    notifications.count { !it.isRead }
                } else {
                    notifications.count { !it.isRead && types.contains(it.type) }
                }
            }
        }
}

@HiltViewModel
class NotificationsViewModel @Inject constructor(
    private val notificationRepository: NotificationRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(NotificationsUiState())
    val uiState: StateFlow<NotificationsUiState> = _uiState.asStateFlow()

    init {
        loadNotifications()
    }

    fun setFilter(filter: NotificationFilter) {
        _uiState.value = _uiState.value.copy(selectedFilter = filter)
    }

    fun loadNotifications() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)

            notificationRepository.getNotifications(page = 1, limit = 50)
                .onSuccess { paginatedData ->
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        notifications = paginatedData.data,
                        currentPage = 1,
                        totalPages = paginatedData.totalPages
                    )
                }
                .onFailure { exception ->
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        error = exception.message ?: "알림을 불러오는데 실패했습니다"
                    )
                }
        }
    }

    fun loadMoreNotifications() {
        val currentState = _uiState.value
        if (currentState.isLoadingMore || !currentState.hasMorePages) return

        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoadingMore = true)

            val nextPage = currentState.currentPage + 1
            notificationRepository.getNotifications(page = nextPage)
                .onSuccess { paginatedData ->
                    _uiState.value = _uiState.value.copy(
                        isLoadingMore = false,
                        notifications = currentState.notifications + paginatedData.data,
                        currentPage = nextPage,
                        totalPages = paginatedData.totalPages
                    )
                }
                .onFailure { exception ->
                    _uiState.value = _uiState.value.copy(
                        isLoadingMore = false,
                        error = exception.message
                    )
                }
        }
    }

    fun markAsRead(notification: AppNotification) {
        if (notification.isRead) return

        viewModelScope.launch {
            notificationRepository.markAsRead(notification.id)
                .onSuccess { updatedNotification ->
                    // Update local state
                    val updatedList = _uiState.value.notifications.map {
                        if (it.id == notification.id) {
                            updatedNotification ?: it.copy(
                                status = com.parkgolf.app.domain.model.NotificationStatus.READ,
                                readAt = java.time.LocalDateTime.now()
                            )
                        } else it
                    }
                    _uiState.value = _uiState.value.copy(
                        notifications = updatedList
                    )
                }
        }
    }

    fun markAllAsRead() {
        viewModelScope.launch {
            notificationRepository.markAllAsRead()
                .onSuccess { count ->
                    // Update all notifications to read
                    val updatedList = _uiState.value.notifications.map {
                        if (!it.isRead) {
                            it.copy(
                                status = com.parkgolf.app.domain.model.NotificationStatus.READ,
                                readAt = java.time.LocalDateTime.now()
                            )
                        } else it
                    }
                    _uiState.value = _uiState.value.copy(
                        notifications = updatedList
                    )
                }
        }
    }

    fun deleteNotification(notification: AppNotification) {
        viewModelScope.launch {
            notificationRepository.deleteNotification(notification.id)
                .onSuccess {
                    val updatedList = _uiState.value.notifications.filter { it.id != notification.id }
                    _uiState.value = _uiState.value.copy(
                        notifications = updatedList
                    )
                }
                .onFailure { exception ->
                    _uiState.value = _uiState.value.copy(
                        error = "알림 삭제에 실패했습니다"
                    )
                }
        }
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(error = null)
    }

    fun refresh() {
        loadNotifications()
    }
}

// Extension function for copying AppNotification (since it's a data class)
private fun AppNotification.copy(
    status: com.parkgolf.app.domain.model.NotificationStatus = this.status,
    readAt: java.time.LocalDateTime? = this.readAt
): AppNotification {
    return AppNotification(
        id = this.id,
        userId = this.userId,
        type = this.type,
        title = this.title,
        message = this.message,
        data = this.data,
        status = status,
        readAt = readAt,
        createdAt = this.createdAt,
        updatedAt = java.time.LocalDateTime.now()
    )
}

File: apps/user-app-ios/Sources/Features/Notifications/NotificationsViewModel.swift
import Foundation

// MARK: - Notification Filter

enum NotificationFilter: String, CaseIterable, Identifiable {
    case all = "전체"
    case booking = "예약"
    case social = "소셜"
    case system = "시스템"

    var id: String { rawValue }

    var icon: String {
        switch self {
        case .all: return "bell.fill"
        case .booking: return "calendar"
        case .social: return "person.2.fill"
        case .system: return "shield.fill"
        }
    }

    var types: [NotificationType]? {
        switch self {
        case .all:
            return nil
        case .booking:
            return [.bookingConfirmed, .bookingCancelled, .paymentSuccess, .paymentFailed]
        case .social:
            return [.friendRequest, .friendAccepted, .chatMessage]
        case .system:
            return [.systemAlert]
        }
    }

    func matches(_ type: NotificationType) -> Bool {
        guard let types = types else { return true }
        return types.contains(type)
    }
}

// MARK: - Notifications ViewModel

@MainActor
final class NotificationsViewModel: ObservableObject {
    // MARK: - Published Properties

    @Published var notifications: [AppNotification] = []
    @Published var isLoading = false
    @Published var isLoadingMore = false
    @Published var error: String?
    @Published var selectedFilter: NotificationFilter = .all

    // MARK: - Computed Properties

    var filteredNotifications: [AppNotification] {
        guard let types = selectedFilter.types else {
            return notifications
        }
        return notifications.filter { types.contains($0.type) }
    }

    var unreadCount: Int {
        notifications.filter { !$0.isRead }.count
    }

    var unreadCounts: [NotificationFilter: Int] {
        var counts: [NotificationFilter: Int] = [:]
        for filter in NotificationFilter.allCases {
            if let types = filter.types {
                counts[filter] = notifications.filter { !$0.isRead && types.contains($0.type) }.count
            } else {
                counts[filter] = notifications.filter { !$0.isRead }.count
            }
        }
        return counts
    }

    func emptyMessage(for filter: NotificationFilter) -> (title: String, description: String) {
        switch filter {
        case .all:
            return ("알림이 없습니다", "새로운 알림이 도착하면 여기에 표시됩니다")
        case .booking:
            return ("예약 알림이 없습니다", "예약 관련 알림이 도착하면 여기에 표시됩니다")
        case .social:
            return ("소셜 알림이 없습니다", "친구 요청이나 채팅 알림이 도착하면 여기에 표시됩니다")
        case .system:
            return ("시스템 알림이 없습니다", "시스템 공지사항이 도착하면 여기에 표시됩니다")
        }
    }

    // MARK: - Pagination

    private var currentPage = 1
    private var totalPages = 1
    private var hasMorePages: Bool { currentPage < totalPages }

    // MARK: - Dependencies

    private let notificationService = NotificationService()

    // MARK: - Public Methods

    func loadNotifications() async {
        guard !isLoading else { return }

        isLoading = true
        error = nil
        currentPage = 1

        do {
            let data = try await notificationService.getNotifications(page: 1, limit: 50)
            notifications = data.notifications
            totalPages = data.totalPages
            currentPage = 1
        } catch {
            self.error = error.localizedDescription
        }

        isLoading = false
    }

    func loadMoreNotifications() async {
        guard !isLoadingMore, hasMorePages else { return }

        isLoadingMore = true
        let nextPage = currentPage + 1

        do {
            let data = try await notificationService.getNotifications(page: nextPage)
            notifications.append(contentsOf: data.notifications)
            totalPages = data.totalPages
            currentPage = nextPage
        } catch {
            self.error = error.localizedDescription
        }

        isLoadingMore = false
    }

    func markAsRead(_ notification: AppNotification) async {
        guard !notification.isRead else { return }

        do {
            _ = try await notificationService.markAsRead(notificationId: notification.id)

            // Update local state
            if let index = notifications.firstIndex(where: { $0.id == notification.id }) {
                let updatedNotification = notifications[index]
                // Create a new notification with updated readAt
                notifications[index] = AppNotification(
                    id: updatedNotification.id,
                    userId: updatedNotification.userId,
                    type: updatedNotification.type,
                    title: updatedNotification.title,
                    message: updatedNotification.message,
                    data: updatedNotification.data,
                    status: .read,
                    readAt: Date(),
                    createdAt: updatedNotification.createdAt,
                    updatedAt: Date()
                )
            }
        } catch {
            self.error = error.localizedDescription
        }
    }

    func markAllAsRead() async {
        do {
            _ = try await notificationService.markAllAsRead()

            // Update local state
            notifications = notifications.map { notification in
                if !notification.isRead {
                    return AppNotification(
                        id: notification.id,
                        userId: notification.userId,
                        type: notification.type,
                        title: notification.title,
                        message: notification.message,
                        data: notification.data,
                        status: .read,
                        readAt: Date(),
                        createdAt: notification.createdAt,
                        updatedAt: Date()
                    )
                }
                return notification
            }
        } catch {
            self.error = error.localizedDescription
        }
    }

    func deleteNotification(_ notification: AppNotification) async {
        do {
            try await notificationService.deleteNotification(notificationId: notification.id)

            // Update local state
            notifications.removeAll { $0.id == notification.id }
        } catch {
            self.error = error.localizedDescription
        }
    }

    func refreshUnreadCount() async {
        do {
            _ = try await notificationService.getUnreadCount()
        } catch {
            // Silently fail for unread count refresh
        }
    }
}
